usr/src/uts/common/fs/zfs/spa_history.c

```diff
@@ -192,6 +195,50 @@ spa_history_zone(void)
 }
 
 /*
+ * Post a history sysevent.  The nvlist_t* passed into this function will be
+ * duplicated (by nvlist_add_nvlist) and should be freed by caller.
+ * This function will only run in the context of the kernel.
+ */
+static void
+spa_history_log_notify(spa_t *spa, nvlist_t *nvl)
```

Comment:
I see the duplication as well and agree this makes sense.

`spa_event_create()` and `spa_event_post()` are both static functions to `spa.c`, with the wrapper function `spa_event_notify()` being accessible outside of the file.  Would it make sense to modify both `spa_event_create()` as well as `spa_event_notify()` to take the addition `nvlist_t*` argument?  It would look something like this.

inside `spa_history.c`

``` c
/*
 * Post a history sysevent.  The nvlist_t* passed into this function will be
 * duplicated and should be freed by caller.
 */
static void
spa_history_log_notify(spa_t *spa, nvlist_t *nvl)
{
        spa_event_notify(spa, NULL, nvl, ESC_ZFS_HISTORY_EVENT);
}
```

inside `spa.c`

``` c
void
spa_event_notify(spa_t *spa, vdev_t *vd, nvlist_t *nvl, const char *name)
{
        spa_event_post(spa_event_create(spa, vd, nvl, name));
}
```

and the relevant bits from `spa_event_create()`

``` c
static sysevent_t *
spa_event_create(spa_t *spa, vdev_t *vd, nvlist_t *nvl, const char *name)
{
        ...
        if (nvl) {
                value.value_type = SE_DATA_TYPE_NVLIST;
                value.value.sv_nvlist = nvl;
                if (sysevent_add_attr(&attr, ZFS_EV_HISTORY, &value,
                    SE_SLEEP) != 0)
                        goto done;
        }
        ...
}
```

Also, `spa_event_create()` saves the argument `nvl` as `ZFS_EV_HISTORY`: maybe it should just be noted that the optional `nvllist_t*` argument will get saved as `ZFS_EV_HISTORY`, or perhaps that string should be an argument as well?